Compatibility-2.0.md

```diff
@@ -26,3 +26,304 @@ Embedded Mode also might use a decorated Cache using a Transcoder. For example t
+ is unmarshalled via a protostream marshaller into a java representation.
+* Caches used in the server with no indexing and compatibility use the ```IdentityEncoder```, which treats content read and written as opaque ```byte[]```.
+* Caches used in the server with compatibility (and no indexing) uses the CompatModeEncoder, which takes the ```byte[]``` from the client and unmarshalls them to java object for storage.
+* Scripts which carry data-type ```utf-8``` wraps a cache with the ```UTF8Encoder```, that will extract ```byte[]``` out of a String in order to store and will create Strings out of ```bytes[]``` for reading
+operations.
+
+
+Some endpoints in the Infinispan Server allow to send and request data in a certain format (MimeType):
+
+* Scripting: scripts (.js at least) can pass a header "dataType" with a MimeType string;
+* Rest: request headers "content-type" (for PUT/POST) and "accept" (for GET) can carry mime types as well
+* HotRod: there is no such capability yet, but it's something that will eventually be added.
+
+
+How does the Mimetype relate to the encoder general mechanism?
+
+
+## Proposal
+
+The proposal is to use a pair (Encoder, MimeType) to describe cache storage. Examples: 
+
+* "```application/json``` with no encoding" means IdentityEncoder associated of a content that represents a json (```byte[]```, ```Gson```, ```Map<String, Map<?,?>>```, etc)
+* "```text/plain; charset=utf-8``` with marshalling" means content produced by the MarshallerEncoder that represents text, so it's a ```byte[]```
+* "```application/octet-stream``` with KryoEncoder" means ```byte[]``` produced by a KryoMarshaller.
+
+This means that different MimeTypes can be associated with the same Encoder, e.g. the IdentityEncoder can be associated with any MimeType meaning whatever is sent
+to cache will be written or read without transformations.
+
+The purpose of the MimeType is twofold:
+
+* Provide a metadata about the content of the cache, independently of the encoding;
+* By describing the format, enable the conversion between different encodings, a.k.a. Transcoding.
+
+
+### Configuration changes
+
+A cache will have a data-type section that looks like the following:
+
+```xml
+<local-cache>
+  <data-type>
+     <key mime-type="application/xml" encoder="org.infinispan.commons.dataconversion.IdentityEncoder"/>
+     <value mime-type="text/plain; charset=UTF-8" encoder="org.infinispan.dataconversion.MarshalledEncoder"/>
+  </data-type>
+</local-cache>
+```
+
+The MimeType will be either present in the config or inferred; if not provided, Infinispan will automatically pick the best 
+decorated cache depending on the situation.
+
+
+### API changes
+Apart from being able to use a different encoders on the fly, it will be possible to ask for a different MimeType using the method:
+
+```java
+// Both keys and types as "application/xml"
+cache.getAdvancedCache.withMimeType("application/xml").withEncoding(IdentityEncoder.class)
```

Comment:
What about charset here? Of course this only kicks in when we try to use mime type "text/plain" or "application/text".